apis/kubedb/v1alpha2/kafka_webhook.go

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright AppsCode Inc. and Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"errors"
+
+	errors2 "github.com/pkg/errors"
+	"gomodules.xyz/pointer"
+	core "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var kafkalog = logf.Log.WithName("kafka-resource")
+
+func (k *Kafka) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(k).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-kafka-kubedb-com-v1alpha1-kafka,mutating=true,failurePolicy=fail,sideEffects=None,groups=kubedb.com,resources=kafkas,verbs=create,versions=v1alpha1,name=mkafka.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Defaulter = &Kafka{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (k *Kafka) Default() {
+	if k == nil {
+		return
+	}
+	kafkalog.Info("default", "name", k.Name)
+	k.SetDefaults()
+}
+
+//+kubebuilder:webhook:path=/validate-kafka-kubedb-com-v1alpha1-kafka,mutating=false,failurePolicy=fail,sideEffects=None,groups=kubedb.com,resources=kafkas,verbs=create;update,versions=v1alpha1,name=vkafka.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Validator = &Kafka{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (k *Kafka) ValidateCreate() error {
+	kafkalog.Info("validate create", "name", k.Name)
+	err := k.ValidateCreateOrUpdate()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (k *Kafka) ValidateUpdate(old runtime.Object) error {
+	err := k.ValidateCreateOrUpdate()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (k *Kafka) ValidateDelete() error {
+	kafkalog.Info("validate delete", "name", k.Name)
+
+	// TODO(user): fill in your validation logic upon object deletion.
+	return nil
+}
+
+func (k *Kafka) ValidateCreateOrUpdate() error {
```

Comment:
We also need to validate storage related stuffs.
Check this https://github.com/kubedb/mongodb/blob/master/pkg/admission/validator.go#L325